qualtran/bloqs/chemistry/thc.ipynb

```diff
@@ -0,0 +1,302 @@
+    "cq.autogen": "_make_uniform_superposition.md"
+   },
+   "source": [
+    "## `UniformSuperpositionTHC`\n",
+    "Prepare uniform superposition state for THC.\n",
+    "\n",
+    "$$\n",
+    "    |0\\rangle^{\\otimes 2\\log(M+1)} \\rightarrow \\sum_{\\mu\\le\\nu}^{M} |\\mu\\rangle|\\nu\\rangle\n",
+    "    + \\sum_{\\mu}^{N/2}|\\mu\\rangle|\\nu=M+1\\rangle,\n",
+    "$$\n",
+    "\n",
+    "where $M$ is the THC auxiliary dimension, and $N$ is the number of spin orbitals.\n",
+    "\n",
+    "The toffoli complexity of this gate should be $10 \\log(M+1) + 2 b_r - 9$.\n",
+    "Currently it is a good deal larger due to:\n",
+    " 1. inverting inequality tests should not need more toffolis.\n",
+    " 2. We are not using phase-gradient gate toffoli cost for Ry rotations\n",
+    " 3. Small differences in quoted vs implemented comparator costs.\n",
+    "\n",
+    "See: https://github.com/quantumlib/Qualtran/issues/390\n",
+    "\n",
+    "#### Parameters\n",
+    " - `num_mu`: THC auxiliary index dimension $M$\n",
+    " - `num_spin_orb`: number of spin orbitals $N$ \n",
+    "\n",
+    "#### Registers\n",
+    " - `mu`: $\\mu$ register.\n",
+    " - `nu`: $\\nu$ register.\n",
+    " - `succ`: ancilla flagging success of amplitude amplification.\n",
+    " - `eq_nu_mp1`: ancillas for flagging if $\\nu = M+1$. \n",
+    "\n",
+    "#### References\n",
+    "[Even more efficient quantum computations of chemistry through tensor hypercontraction](https://arxiv.org/pdf/2011.03494.pdf). Eq. 29.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1b6bb0e0",
+   "metadata": {
+    "cq.autogen": "_make_uniform_superposition.py"
+   },
+   "outputs": [],
+   "source": [
+    "from qualtran.bloqs.chemistry.thc import UniformSuperpositionTHC\n",
+    "\n",
+    "num_mu = 10\n",
+    "num_spin_orb = 4\n",
+    "bloq = UniformSuperpositionTHC(num_mu=num_mu, num_spin_orb=num_spin_orb)\n",
+    "show_bloq(bloq)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a89e358f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from qualtran.resource_counting import get_bloq_counts_graph\n",
+    "from qualtran.drawing import GraphvizCounts\n",
+    "from qualtran.bloqs.chemistry.thc_notebook_utils import generalize\n",
+    "graph, sigma = get_bloq_counts_graph(bloq, generalizer=generalize)\n",
+    "GraphvizCounts(graph).get_svg()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f2cbc8e6",
+   "metadata": {},
+   "source": [
+    "Let's print out the costs contributing to the TCount."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "fa6858a5",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from qualtran.bloqs.chemistry.thc_notebook_utils import bin_bloq_counts\n",
+    "\n",
+    "binned_counts = bin_bloq_counts(bloq)\n",
+    "# number of bits for mu register (nm in THC paper)\n",
+    "nm = bloq.signature[0].bitsize\n",
```

Comment:
use `bloq.num_mu`. Relying on `signature[0]` to be what you want it to be is dangerous!

Reply:
Done.